diff --color -urpN a/include/linux/socket.h b/include/linux/socket.h
--- a/include/linux/socket.h	2024-03-10 20:38:09.000000000 +0000
+++ b/include/linux/socket.h	2024-12-16 15:17:15.480582610 +0000
@@ -240,7 +240,8 @@ struct ucred {
 				 * transport protocol
 				 */
 
-#define AF_MAX		46	/* For now.. */
+#define AF_MAX		47	/* For now.. */
+#define AF_RAKAIA		46	/* RAKAIA implementation */
 
 /* Protocol families, same as address families. */
 #define PF_UNSPEC	AF_UNSPEC
@@ -292,6 +293,7 @@ struct ucred {
 #define PF_XDP		AF_XDP
 #define PF_MCTP		AF_MCTP
 #define PF_MAX		AF_MAX
+#define PF_RAKAIA		AF_RAKAIA
 
 /* Maximum queue length specifiable by listen.  */
 #define SOMAXCONN	4096
@@ -384,6 +386,7 @@ struct ucred {
 #define SOL_MCTP	285
 #define SOL_SMC		286
 #define SOL_VSOCK	287
+#define SOL_RAKAIA	288
 
 /* IPX options */
 #define IPX_TYPE	1
diff --color -urpN a/include/uapi/linux/rakaia.h b/include/uapi/linux/rakaia.h
--- a/include/uapi/linux/rakaia.h	1970-01-01 00:00:00.000000000 +0000
+++ b/include/uapi/linux/rakaia.h	2024-12-16 15:17:15.480582610 +0000
@@ -0,0 +1,40 @@
+/* SPDX-License-Identifier: GPL-2.0 WITH Linux-syscall-note */
+/*
+ * Kernel Connection Multiplexor
+ *
+ * Copyright (c) 2016 Tom Herbert <tom@herbertland.com>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2
+ * as published by the Free Software Foundation.
+ *
+ * User API to clone KCM sockets and attach transport socket to a KCM
+ * multiplexor.
+ */
+
+#ifndef RAKAIA_KERNEL_H
+#define RAKAIA_KERNEL_H
+
+struct rakaia_attach {
+	int fd;
+	int bpf_fd;
+};
+
+struct rakaia_unattach {
+	int fd;
+};
+
+struct rakaia_clone {
+	int fd;
+};
+
+#define SIOCRAKAIAATTACH (SIOCPROTOPRIVATE + 0)
+#define SIOCRAKAIAUNATTACH (SIOCPROTOPRIVATE + 1)
+#define SIOCRAKAIACLONE (SIOCPROTOPRIVATE + 2)
+
+#define RAKAIAPROTO_CONNECTED 0
+
+/* Socket options */
+#define RAKAIA_RECV_DISABLE 1
+
+#endif
diff --color -urpN a/Makefile b/Makefile
--- a/Makefile	2024-03-10 20:38:09.000000000 +0000
+++ b/Makefile	2024-12-16 15:17:15.480582610 +0000
@@ -2,7 +2,7 @@
 VERSION = 6
 PATCHLEVEL = 8
 SUBLEVEL = 0
-EXTRAVERSION =
+EXTRAVERSION = -rakaia
 NAME = Hurr durr I'ma ninja sloth
 
 # *DOCUMENTATION*
diff --color -urpN a/net/ipv4/tcp.c b/net/ipv4/tcp.c
--- a/net/ipv4/tcp.c	2024-03-10 20:38:09.000000000 +0000
+++ b/net/ipv4/tcp.c	2024-12-16 15:17:50.337153269 +0000
@@ -653,6 +653,7 @@ void tcp_mark_push(struct tcp_sock *tp,
 	TCP_SKB_CB(skb)->tcp_flags |= TCPHDR_PSH;
 	tp->pushed_seq = tp->write_seq;
 }
+EXPORT_SYMBOL(tcp_mark_push);
 
 static inline bool forced_push(const struct tcp_sock *tp)
 {
diff --color -urpN a/net/ipv4/tcp_output.c b/net/ipv4/tcp_output.c
--- a/net/ipv4/tcp_output.c	2024-03-10 20:38:09.000000000 +0000
+++ b/net/ipv4/tcp_output.c	2024-12-16 15:17:59.357042421 +0000
@@ -1879,6 +1879,7 @@ unsigned int tcp_current_mss(struct sock
 
 	return mss_now;
 }
+EXPORT_SYMBOL(tcp_current_mss);
 
 /* RFC2861, slow part. Adjust cwnd, after it was not full during one rto.
  * As additional protections, we do not touch cwnd in retransmission phases,
@@ -2682,6 +2683,7 @@ void tcp_chrono_stop(struct sock *sk, co
 	else if (type == tp->chrono_type)
 		tcp_chrono_set(tp, TCP_CHRONO_BUSY);
 }
+EXPORT_SYMBOL(tcp_chrono_stop);
 
 /* This routine writes packets to the network.  It advances the
  * send_head.  This happens as incoming acks open up the remote
@@ -2978,6 +2980,8 @@ void __tcp_push_pending_frames(struct so
 			   sk_gfp_mask(sk, GFP_ATOMIC)))
 		tcp_check_probe_timer(sk);
 }
+EXPORT_SYMBOL(__tcp_push_pending_frames);
+
 
 /* Send _single_ skb sitting at the send head. This function requires
  * true push pending frames to setup probe timer etc.
@@ -3530,6 +3534,7 @@ void sk_forced_mem_schedule(struct sock
 		mem_cgroup_charge_skmem(sk->sk_memcg, amt,
 					gfp_memcg_charge() | __GFP_NOFAIL);
 }
+EXPORT_SYMBOL(sk_forced_mem_schedule);
 
 /* Send a FIN. The caller locks the socket for us.
  * We should try to send a FIN packet really hard, but eventually give up.
diff --color -urpN a/net/kcm/kcmsock.c b/net/kcm/kcmsock.c
--- a/net/kcm/kcmsock.c	2024-03-10 20:38:09.000000000 +0000
+++ b/net/kcm/kcmsock.c	2024-12-16 15:17:15.480582610 +0000
@@ -1321,7 +1321,7 @@ static int kcm_attach_ioctl(struct socke
 	if (!csock)
 		return -ENOENT;
 
-	prog = bpf_prog_get_type(info->bpf_fd, BPF_PROG_TYPE_SOCKET_FILTER);
+	prog = bpf_prog_get_type(info->bpf_fd, BPF_PROG_TYPE_SK_SKB);
 	if (IS_ERR(prog)) {
 		err = PTR_ERR(prog);
 		goto out;
diff --color -urpN a/net/socket.c b/net/socket.c
--- a/net/socket.c	2024-03-10 20:38:09.000000000 +0000
+++ b/net/socket.c	2024-12-16 15:17:15.480582610 +0000
@@ -218,6 +218,7 @@ static const char * const pf_family_name
 	[PF_SMC]	= "PF_SMC",
 	[PF_XDP]	= "PF_XDP",
 	[PF_MCTP]	= "PF_MCTP",
+	[PF_RAKAIA]	= "PF_RAKAIA",
 };
 
 /*
diff --color -urpN a/net/strparser/strparser.c b/net/strparser/strparser.c
--- a/net/strparser/strparser.c	2024-03-10 20:38:09.000000000 +0000
+++ b/net/strparser/strparser.c	2024-12-16 15:17:15.481582598 +0000
@@ -209,7 +209,7 @@ static int __strp_recv(read_descriptor_t
 
 		if (!stm->strp.full_len) {
 			ssize_t len;
-
+            *((int *)head->cb) = stm->strp.offset;
 			len = (*strp->cb.parse_msg)(strp, head);
 
 			if (!len) {
diff --color -urpN a/security/selinux/hooks.c b/security/selinux/hooks.c
--- a/security/selinux/hooks.c	2024-03-10 20:38:09.000000000 +0000
+++ b/security/selinux/hooks.c	2024-12-16 15:17:15.481582598 +0000
@@ -1297,7 +1297,7 @@ static inline u16 socket_type_to_securit
 			return SECCLASS_XDP_SOCKET;
 		case PF_MCTP:
 			return SECCLASS_MCTP_SOCKET;
-#if PF_MAX > 46
+#if PF_MAX > 47
 #error New address family defined, please update this function.
 #endif
 		}
diff --color -urpN a/security/selinux/include/classmap.h b/security/selinux/include/classmap.h
--- a/security/selinux/include/classmap.h	2024-03-10 20:38:09.000000000 +0000
+++ b/security/selinux/include/classmap.h	2024-12-16 15:17:15.481582598 +0000
@@ -164,6 +164,7 @@ const struct security_class_mapping secc
 	{ "nfc_socket", { COMMON_SOCK_PERMS, NULL } },
 	{ "vsock_socket", { COMMON_SOCK_PERMS, NULL } },
 	{ "kcm_socket", { COMMON_SOCK_PERMS, NULL } },
+	{ "rakaia_socket", { COMMON_SOCK_PERMS, NULL } },
 	{ "qipcrtr_socket", { COMMON_SOCK_PERMS, NULL } },
 	{ "smc_socket", { COMMON_SOCK_PERMS, NULL } },
 	{ "infiniband_pkey", { "access", NULL } },
@@ -181,6 +182,6 @@ const struct security_class_mapping secc
 	{ NULL }
 };
 
-#if PF_MAX > 46
+#if PF_MAX > 47
 #error New address family defined, please update secclass_map.
 #endif
